src/mpid/common/genq/mpidu_genq_shmem_queue.h

```diff
@@ -83,103 +155,92 @@ static inline MPIDU_genqi_shmem_cell_header_s
 {
     int rc = MPI_SUCCESS;
-    MPIDU_genqi_shmem_cell_header_s *cell_h = CELL_TO_HEADER(cell);
+    genqi_shmem_cell_header_s *cell_h = CELL_TO_HEADER(cell);
+    cell_h->u.inverse_queue.next = 0;
+    cell_h->u.inverse_queue.prev = 0;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
+    void *prev_handle = NULL;
+    void *handle = (void *) cell_h->handle;
 
-    if (queue->q.flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
-        uintptr_t handle = cell_h->handle;
+    do {
+        prev_handle = MPL_atomic_load_ptr(&queue->q.tail.m);
+        cell_h->u.inverse_queue.prev = (uintptr_t) prev_handle;
+    } while (MPL_atomic_cas_ptr(&queue->q.tail.m, prev_handle, handle) != prev_handle);
 
-        if (queue->q.tail.s) {
-            HANDLE_TO_HEADER(queue->q.pool, queue->q.tail.s)->next = handle;
-        }
-        cell_h->prev = queue->q.tail.s;
-        queue->q.tail.s = handle;
-        if (!queue->q.head.s) {
-            queue->q.head.s = queue->q.tail.s;
-        }
-    } else {    /* MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC */
-        void *prev_handle = NULL;
-        void *handle = (void *) cell_h->handle;
-
-        do {
-            prev_handle = MPL_atomic_load_ptr(&queue->q.tail.m);
-            cell_h->prev = (uintptr_t) prev_handle;
-        } while (MPL_atomic_cas_ptr(&queue->q.tail.m, prev_handle, handle) != prev_handle);
-    }
-
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
     return rc;
 }
 
-static inline void *MPIDU_genq_shmem_queue_head(MPIDU_genq_shmem_queue_t queue)
-{
-    MPIDU_genqi_shmem_cell_header_s *cell_h = NULL;
+/* EXTERNAL INTERFACE */
 
-    if (queue->q.flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
-        cell_h = HANDLE_TO_HEADER(queue->q.pool, queue->q.head.s);
-    } else {    /* MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC */
-        if (!queue->q.head.s) {
-            cell_h = MPIDU_genqi_shmem_get_head_cell_header(queue);
-        } else {
-            cell_h = HANDLE_TO_HEADER(queue->q.pool, queue->q.head.s);
-        }
-
-    }
-    if (cell_h) {
-        return HEADER_TO_CELL(cell_h);
+static inline int MPIDU_genq_shmem_queue_dequeue(MPIDU_genq_shmem_queues_t queues,
+                                                 int queue_idx, void **cell)
+{
+    int rc = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_INIT);
+
+    genqi_shmem_queue_u *queue_obj = (genqi_shmem_queue_u *) queues.ptr + queue_idx;
+    int flags = queue_obj->q.flags;
+    if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
+        rc = genqi_serial_dequeue(queues.base, queue_obj, cell);
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__INV_MPSC) {
+        rc = genqi_inv_mpsc_dequeue(queues.base, queue_obj, cell);
```

Comment:
`queues.base` was not initialized in `queues_alloc`.

Reply:
Wait for my last commit...